node/network/pov-distribution/src/lib.rs

```diff
@@ -394,35 +430,8 @@ async fn handle_fetch(
 				return;
 			}
 			Entry::Vacant(e) => {
-				if let Ok(Some(relevant_validators)) = determine_relevant_validators(
-					ctx,
-					relay_parent,
-					descriptor.para_id,
-				).await {
-					// We only need one connection request per (relay_parent, para_id)
-					// so here we take this shortcut to avoid calling `connect_to_validators`
-					// more than once.
-					if !state.connection_requests.contains_request(&relay_parent) {
-						match validator_discovery::connect_to_validators(
-							ctx,
-							relay_parent,
-							relevant_validators.clone(),
-						).await {
-							Ok(new_connection_request) => {
-								state.connection_requests.put(relay_parent, new_connection_request);
-							}
-							Err(e) => {
-								tracing::debug!(
-									target: LOG_TARGET,
-									"Failed to create a validator connection request {:?}",
-									e,
-								);
-							}
-						}
-					}
-
-					e.insert(vec![response_sender]);
-				}
+				connect_to_relevant_validators(&mut state.connection_requests, ctx, relay_parent, &descriptor).await;
+				e.insert(vec![response_sender]);
```

Comment:
I think we should not insert the sender unconditionally. If there are no validators to connect to, the receiver will be pending forever.

Reply:
I'm not sure about this. AFAIK a runtime request can fail spuriously while we can be already connected to the relevant peers.
And we will clean it up when the relay parent becomes irrelevant.

Reply:
The previous check did not check whether we were able to connect to validators, but whether there are any relevant validators.

Reply:
I think there should always be relevant validators in theory.
However, the caller should anyway handle the case that we issued the connection request but haven't received the PoV for a while. And it does this by issuing a separate job for it in the background that should not block progress on other jobs: https://github.com/paritytech/polkadot/blob/57eb9d10d874a27f2661523ce4b533aa690589c3/node/core/backing/src/lib.rs#L568.

Reply:
What I'm saying is yes, I've changed the behavior here, but it is equally "correct" as the previous one.

Reply:
Just wanted to double check, the changed code sure looks nicer ;-)